src/pie_models/models/token_classification_with_seq2seq_encoder_and_crf.py

```diff
@@ -0,0 +1,202 @@
+from torchcrf import CRF
+from transformers import AutoConfig, AutoModel, BatchEncoding
+from transformers.modeling_outputs import TokenClassifierOutput
+from typing_extensions import TypeAlias
+
+from .components.seq2seq_encoder import build_seq2seq_encoder
+
+ModelBatchEncodingType: TypeAlias = BatchEncoding
+ModelBatchOutputType: TypeAlias = Dict[str, Any]
+
+StepBatchEncodingType: TypeAlias = Tuple[
+    Dict[str, Tensor],
+    Optional[Tensor],
+]
+
+HF_MODEL_TYPE_TO_CLASSIFIER_DROPOUT_ATTRIBUTE = {
+    "albert": "classifier_dropout_prob",
+    "distilbert": "seq_classif_dropout",
+    "longformer": "hidden_dropout_prob"
+}
+
+TRAINING = "train"
+VALIDATION = "val"
+TEST = "test"
+
+
+@PyTorchIEModel.register()
+class TokenClassificationModelWithSeq2SeqEncoderAndCrf(PyTorchIEModel):
+    def __init__(
+        self,
+        model_name_or_path: str,
+        num_classes: int,
+        learning_rate: float = 1e-5,
+        task_learning_rate: Optional[float] = None,
+        label_pad_token_id: int = -100,
+        ignore_index: int = 0,
+        classifier_dropout: Optional[float] = None,
+        use_crf: bool = True,
+        freeze_base_model: bool = False,
+        seq2seq_encoder: Optional[Dict[str, Any]] = None,
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.save_hyperparameters()
+
+        self.ignore_index = ignore_index
+
+        self.learning_rate = learning_rate
+        self.task_learning_rate = task_learning_rate
+        self.label_pad_token_id = label_pad_token_id
+        self.num_classes = num_classes
+
+        config = AutoConfig.from_pretrained(model_name_or_path)
+        if self.is_from_pretrained:
+            self.model = AutoModel.from_config(config=config)
+        else:
+            self.model = AutoModel.from_pretrained(model_name_or_path, config=config)
+
+        if freeze_base_model:
+            self.model.requires_grad_(False)
+
+        hidden_size = config.hidden_size
+        self.seq2seq_encoder = None
+        if seq2seq_encoder is not None:
+            self.seq2seq_encoder, hidden_size = build_seq2seq_encoder(
+                config=seq2seq_encoder, input_size=hidden_size
+            )
+
+        # Get the classifier dropout value from the Huggingface model config.
+        # This is a bit of a mess since some Configs use different variable names or change the semantics
+        # of the dropout (e.g. DistilBert has one dropout prob for QA and one for Seq classification, and a
+        # general one for embeddings, encoder and pooler).
+        classifier_dropout_attr = HF_MODEL_TYPE_TO_CLASSIFIER_DROPOUT_ATTRIBUTE.get(
+            config.model_type, "classifier_dropout"
+        )
+        if classifier_dropout is None:
+            classifier_dropout = getattr(config, classifier_dropout_attr) or 0.0
+        else:
+            raise ValueError(f"The config {type(config),__name__} loaded from {model_name_or_path} has no attribute {classifier_dropout_attr}")
```

Comment:
It does not make sense to raise the exception in this case. It should happen if `not hasattr(config, classifier_dropout_attr)` (and only if `classifier_dropout` is not provided).

Reply:
Which one?

Reply:
Oh my bad I haven't pushed it yet